File: 3734-lexicographically-smallest-palindromic-permutation-greater-than-target/3734-lexicographically-smallest-palindromic-permutation-greater-than-target.cpp
class Solution {
public:
    string lexPalindromicPermutation(string s, string target) {
        int n = s.length();
        vector<int> freq(26, 0);
        for (char c : s) {
            freq[c - 'a']++;
        }

        // 1. Check palindrome feasibility & build half-frequency map
        int odd_count = 0;
        char odd_char = 0;
        vector<int> half_freq(26, 0);

        for (int i = 0; i < 26; ++i) {
            if (freq[i] % 2 != 0) {
                odd_count++;
                odd_char = 'a' + i;
            }
            half_freq[i] = freq[i] / 2;
        }

        if ((n % 2 == 0 && odd_count != 0) || (n % 2 == 1 && odd_count != 1)) {
            return "";
        }

        int m = n / 2;

        // Helper lambda to construct full palindrome from half string
        auto make_pal = [&](const string& half) -> string {
            string res = half;
            if (n % 2 == 1) res += odd_char;
            for (int i = m - 1; i >= 0; --i) {
                res += half[i];
            }
            return res;
        };

        string ans = "";

        // 2. Case A: Try matching target[0...m-1] exactly
        vector<int> prefix_freq(26, 0);
        bool can_exact = true;
        for (int i = 0; i < m; ++i) {
            prefix_freq[target[i] - 'a']++;
        }
        for (int c = 0; c < 26; ++c) {
            if (half_freq[c] < prefix_freq[c]) {
                can_exact = false;
                break;
            }
        }

        if (can_exact) {
            string exact_half = target.substr(0, m);
            string cand = make_pal(exact_half);
            if (cand > target) {
                ans = cand;
            }
        }

        // 3. Case B: Backtrack to find the rightmost index i where a larger char can be placed
        for (int i = m - 1; i >= 0; --i) {
            // Recalculate prefix frequency for target[0...i-1]
            vector<int> p_freq(26, 0);
            for (int j = 0; j < i; ++j) {
                p_freq[target[j] - 'a']++;
            }

            bool valid_prefix = true;
            for (int c = 0; c < 26; ++c) {
                if (half_freq[c] < p_freq[c]) {
                    valid_prefix = false;
                    break;
                }
            }
            if (!valid_prefix) continue;

            // Remaining available characters for index i and beyond
            vector<int> rem_freq(26, 0);
            for (int c = 0; c < 26; ++c) {
                rem_freq[c] = half_freq[c] - p_freq[c];
            }

            // Find smallest character strictly greater than target[i]
            for (int c = target[i] - 'a' + 1; c < 26; ++c) {
                if (rem_freq[c] > 0) {
                    rem_freq[c]--;

                    // Build greedy minimal tail for positions i+1 ... m-1
                    string half = target.substr(0, i);
                    half += (char)('a' + c);
                    for (int k = 0; k < 26; ++k) {
                        half.append(rem_freq[k], 'a' + k);
                    }

                    string cand = make_pal(half);
                    if (ans.empty() || cand < ans) {
                        ans = cand;
                    }
                    break; // Found smallest larger char for this position i
                }
            }
        }

        return ans;
    }
};

File: 3734-lexicographically-smallest-palindromic-permutation-greater-than-target/3734-lexicographically-smallest-palindromic-permutation-greater-than-target.py
from collections import Counter

class Solution:
    def lexPalindromicPermutation(self, s: str, target: str) -> str:
        n = len(s)
        freq = Counter(s)
        
        odd_char = ""
        odd_count = 0
        half_freq = Counter()
        
        for ch, count in freq.items():
            if count % 2 != 0:
                odd_count += 1
                odd_char = ch
            half_freq[ch] = count // 2
            
        if (n % 2 == 0 and odd_count != 0) or (n % 2 == 1 and odd_count != 1):
            return ""
            
        m = n // 2

        def make_pal(half: str) -> str:
            return half + odd_char + half[::-1]

        ans = None

        prefix_freq = Counter(target[:m])
        if all(half_freq[ch] >= prefix_freq[ch] for ch in prefix_freq):
            exact_half = target[:m]
            cand = make_pal(exact_half)
            if cand > target:
                ans = cand

        for i in range(m - 1, -1, -1):
            prefix_freq = Counter(target[:i])
            if any(half_freq[ch] < prefix_freq[ch] for ch in prefix_freq):
                continue
                
            rem_freq = half_freq - prefix_freq
 
            for c in sorted(rem_freq.keys()):
                if c > target[i] and rem_freq[c] > 0:
                    rem_freq[c] -= 1
        
                    tail = "".join(ch * rem_freq[ch] for ch in sorted(rem_freq.keys()))
                    half = target[:i] + c + tail
                    cand = make_pal(half)
                    
                    if ans is None or cand < ans:
                        ans = cand
                    break 

        return ans if ans is not None else ""